### PR TITLE
[BEAM-8949] SpannerIO integration tests

### PR DESCRIPTION
**Pull request by mszb**
Added Spanner io integration test.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by aaltay** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by mszb** [comment]
@iemejia @aaltay. Could you please run the post-commit tests as well! 

`Run Python PostCommit`

Thanks!

**Comment by iemejia** [comment]
Run Python PostCommit

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
Run Python PostCommit

**Comment by aaltay** [comment]
Run Python 3.5 PostCommit

**Comment by iemejia** [comment]
retest this please

**Comment by lukecwik** [comment]
Run Python PostCommit

**Comment by lukecwik** [comment]
Run Python 3.5 PostCommit

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
mmmmmm

**Comment by iemejia** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
@mszb - is this ready for a review?

**Comment by mszb** [comment]
> @mszb - is this ready for a review?

@aaltay just waiting for all test results. Can you please trigger the tests and the python 3.5 postcommit test!

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
Run Python 3.5 PostCommit

**Comment by iemejia** [comment]
Run Python 3.5 PostCommit

**Comment by iemejia** [comment]
Run Python PostCommit

**Comment by iemejia** [comment]
Run Python 3.5 PostCommit

**Comment by mszb** [comment]
@iemejia thanks for triggering the tests.. Could you please rerun the precommit test `Run Python PreCommit` as it was failed due to wordcount streaming test.

**Comment by iemejia** [comment]
Run Python PreCommit

**Comment by iemejia** [comment]
Run Python PreCommit

**Comment by iemejia** [comment]
Run Python PreCommit

**Comment by iemejia** [comment]
Run Python PreCommit

**Comment by iemejia** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by mszb** [comment]
@aaltay Percommit task failed due to `PortableRunnerTestWithExternalEnv` test. Do you suggest I should rebase the branch again?

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py [hunk 1]
```diff
@@ -1008,31 +1007,30 @@ def _reset_count(self):
     self._cells = 0
 
   def process(self, element):
-    mg_info = element.info
+    for elem in element:
```
Comment (mszb) [inline sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1010]: I've updated the batch process to support the dataflow runner. The process is almost the same as the previous commit but now I've added the `ToList()` combine transform before passing the mutation groups from to the `_BatchFn` to create batches.

**Comment by mszb** [comment]
@chamikaramj have you got any chance to review this pr? Once you’ve done the initial review, ill rebase the branch with master! Ive check the precommit logs and verified that the spanner tests ran successfully.

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
Run Python 3.5 PostCommit

**Comment by iemejia** [comment]
Run Python 3.5 PostCommit

**Comment by mszb** [comment]
@iemejia thanks for triggering the test. I Just review the pre-commit logs that wordcount streaming tests are failed. I've verified them locally and those test cases are passed. Would you please retriggered the precommit test cases! - Thanks

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
@chamikaramj - could you review this PR?

**Comment by mszb** [comment]
Ping @chamikaramj !

**Comment by pabloem** [comment]
I'll take a look next week. Thanks!

**Comment by chamikaramj** [comment]
Thanks Shoaib and Pablo.

**Comment by aaltay** [comment]
@mszb - Do you know why the test is failing?

**Comment by mszb** [comment]
@aaltay - Wordcount stream tests are failing. They are running successfully on my local machine. Jira ticket is already created for that - https://issues.apache.org/jira/browse/BEAM-9767
Maybe retriggering the pre-commit test works but can't say that would pass 100% bcoz of the flaky wordcount test.

Apart from that, there were 3 pre-commit jobs are triggered almost the same time on Jenkins (with ~30seconds difference) and one of them is successfully completed but sadly the failed one is linked with the github.

successful pre-commit: https://builds.apache.org/job/beam_PreCommit_Python_Commit/12541/

failed jobs:
https://builds.apache.org/job/beam_PreCommit_Python_Commit/12542/
https://builds.apache.org/job/beam_PreCommit_Python_Commit/12543/

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
> @aaltay - Wordcount stream tests are failing. They are running successfully on my local machine. Jira ticket is already created for that - https://issues.apache.org/jira/browse/BEAM-9767
> Maybe retriggering the pre-commit test works but can't say that would pass 100% bcoz of the flaky wordcount test.
> 
> Apart from that, there were 3 pre-commit jobs are triggered almost the same time on Jenkins (with ~30seconds difference) and one of them is successfully completed but sadly the failed one is linked with the github.
> 
> successful pre-commit: https://builds.apache.org/job/beam_PreCommit_Python_Commit/12541/
> 
> failed jobs:
> https://builds.apache.org/job/beam_PreCommit_Python_Commit/12542/
> https://builds.apache.org/job/beam_PreCommit_Python_Commit/12543/

OK, let's re-run to de-flake.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py [hunk 2]
```diff
@@ -499,6 +499,7 @@ def test_batch_byte_size(
       # and each bach should contains 25 mutations.
       res = (
           p | beam.Create(mutation_group)
+          | 'combine to list' >> beam.combiners.ToList()
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py:502]: Why do we need to perform this combining to run the test ?

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py [hunk 3]
```diff
@@ -1008,31 +1007,30 @@ def _reset_count(self):
...
+      mg_info = elem.info
+      if mg_info['byte_size'] + self._size_in_bytes > \
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1012]: This seems like a change to the implementation not part of the implementation. Probably should be a separate PR with a JIRA.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 1 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1010]: Can you clarify ? Would would Dataflow need the elements to be combined to a list ? All runners should be able to operate on a PCollection of mutation groups.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py (hunk 2 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py:502]: Yes, the `_BatchFn` requires a single iterable of collection and loop through them to make the batches. Just replicating the same pipeline for the batching in the `_WriteGroup` transform.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 3 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1012]: Sure. Should I create a new Jira ticket and (1) add ticket number in this PR for reference OR (2) create a new PR for this change, and once it gets merge then I rebase this PR and request review? 

I think the first approach required less time to close the tickets! What you suggest?

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 1 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1010]: "Which causes finish_bundle to be called multiple times" do you mean that finish_bundle will be called once per bundle ? 
This is the expected behavior and users will observe this behavior as well. Implementation should work for arbitrary bundle sizes without users having to group PCollection elements together.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py (hunk 2 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py:502]: Do users have to do this as well ? Seems like we are missing something in the implementation. How does Java implementation operate ?

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 3 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1012]: I think (2) is better but we should fix the Spanner connector implementation to work for arbitrary bundle sizes than reducing the bundle to a single element for the test.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py (hunk 2 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py:502]: The user does not have to add ToList transform in the production pipeline. I only added this to test the batch process.
The previous implementation of batching (without ToList transform) was as per the java implementation but without the sorting of the transactions by table and primary key (this is also documented as a feature to be added later).

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 1 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1010]: Make sense, in that case, we don't need to alter the connector code anymore, it was working as expected. Thanks, @chamikaramj for the feedback as it is always helpful.
I'll remove the changes from the spanner io connector and update the IT test code for the assertion.

Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py:1010]: Thanks. Lemme know when this is ready for another look. Also lets trigger the IT with new changes to make sure it passes.

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 3 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py]: Since i've reverted the changes from the connector, there is no need to create new tickets. Resolving this conversation! - Thanks!

**Comment by mszb** [comment]
@chamikaramj: i've made some changes, could you please trigger the tests! 
-Thanks

File: sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py (hunk 2 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio_test.py]: Done!

File: sdks/python/apache_beam/io/gcp/experimental/spannerio.py (hunk 1 above)
Comment (mszb) [reply sdks/python/apache_beam/io/gcp/experimental/spannerio.py]: Done!

**Comment by chamikaramj** [comment]
Thanks. LGTM.

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Run Python PostCommit

**Comment by chamikaramj** [comment]
Run Python PostCommit

**Comment by chamikaramj** [comment]
Run Python PostCommit

**Comment by chamikaramj** [comment]
Run Python PostCommit

**Comment by chamikaramj** [comment]
Run Python 3.7 PostCommit

**Comment by udim** [comment]
Run Python 3.7 PostCommit

**Comment by aaltay** [comment]
Run Python 3.7 PostCommit

**Comment by mszb** [comment]
There were three post-commit jobs triggered for this PR, and the first one is passed.

https://builds.apache.org/job/beam_PostCommit_Python37_PR/144/ (PASSED)
https://builds.apache.org/job/beam_PostCommit_Python37_PR/145/ (Bigquery test failed)
https://builds.apache.org/job/beam_PostCommit_Python37_PR/146/ (Bigquery test failed)

should we try to re-run the job?

**Comment by pabloem** [comment]
it seems that BQ tests were broken. Can you rebase the PR, and we'll rerun the tests?

**Comment by mszb** [comment]
> it seems that BQ tests were broken. Can you rebase the PR, and we'll rerun the tests?

done!

**Comment by aaltay** [comment]
retest this please

**Comment by pabloem** [comment]
retest this plesae

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit

**Comment by pabloem** [comment]
Run Python 3.7 PostCommit
